pkg/restapi/device/operations.go

```diff
@@ -0,0 +1,214 @@
+	return []common.Handler{
+		common.NewHTTPHandler(registerBeginPath, http.MethodGet, o.beginRegistration),
+		common.NewHTTPHandler(registerFinishPath, http.MethodGet, o.finishRegistration),
+	}
+}
+
+func (o *Operation) beginRegistration(w http.ResponseWriter, r *http.Request) {
+	logger.Debugf("handling device registration: %s", r.URL.String())
+
+	userSub := r.URL.Query().Get("user")
+	if userSub == "" {
+		common.WriteErrorResponsef(w, logger, http.StatusBadRequest, "missing user parameter")
+
+		return
+	}
+
+	userData, err := o.store.users.Get(userSub)
+	if err != nil {
+		common.WriteErrorResponsef(w, logger,
+			http.StatusInternalServerError, "failed to read user session cookie: %s", err.Error())
+
+		return
+	}
+
+    device := NewDevice(userData)
+
+	webAuthnUser := protocol.UserEntity{
+		ID:           device.WebAuthnID(),
+		DisplayName:  device.WebAuthnDisplayName(),
+		CredentialEntity: protocol.CredentialEntity{
+			Name:  device.WebAuthnName(),
+		},
+	}
+
+	registerOptions := func(credCreationOpts *protocol.PublicKeyCredentialCreationOptions) {
+		credCreationOpts.User = webAuthnUser
+	}
+
+	// generate PublicKeyCredentialCreationOptions, session data
+	_, sessionData, err := o.webauthn.BeginRegistration(
+		device,
+		registerOptions,
+	)
+
+	if err != nil {
+		common.WriteErrorResponsef(w, logger,
+			http.StatusInternalServerError, "failed to begin registration: %s", err.Error())
+
+		return
+	}
+
+	// store session data as marshaled JSON
+	err = o.sessionStore.SaveWebauthnSession(userData.Sub, sessionData, r, w)
+	if err != nil {
+		common.WriteErrorResponsef(w, logger,
+			http.StatusInternalServerError, "failed to begin registration: %s", err.Error())
+
+		return
+	}
+
+	http.Redirect(w, r, o.uiEndpoint, http.StatusOK)
+	logger.Debugf("Registration begins: %s", o.uiEndpoint)
+}
+
+func (o *Operation) finishRegistration(w http.ResponseWriter, r *http.Request) {
+	logger.Debugf("handling finishing device registration: %s", r.URL.String())
+
+	// todo look for the option to read from the access token,
+	userSub := r.URL.Query().Get("user")
+	if userSub == "" {
+		common.WriteErrorResponsef(w, logger, http.StatusBadRequest, "missing user parameter")
+
+		return
+	}
+
+	userData, err := o.store.users.Get(userSub)
+	if err != nil {
+		common.WriteErrorResponsef(w, logger,
+			http.StatusInternalServerError, "failed to read user session cookie: %s", err.Error())
+
+		return
+	}
+
+	sessionData, err := o.sessionStore.GetWebauthnSession(userData.Sub, r)
+	if err != nil {
+		common.WriteErrorResponsef(w, logger,
+			http.StatusInternalServerError, "failed to get web auth session: %s", err.Error())
+
+		return
+	}
+
+	device := NewDevice(userData)
+
+	credential, err := o.webauthn.FinishRegistration(device, sessionData, r)
+	if err != nil {
+		common.WriteErrorResponsef(w, logger,
+			http.StatusInternalServerError, "failed to finish registration: %s", err.Error())
+
+		return
+	}
+
+	device.AddCredential(*credential)
```

Comment:
The model relationships should be:

* 1:N between user:device
* 1:1 between device:credential

The device authenticator must authenticate with the same credential that is scoped to the website's origin (default), or the [`rpId`](https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredentialRequestOptions/rpId).

Please remove method `AddCredential` - there should only be one credential per device.

Reply:
done